plugins/oauth/server/rest.py

```diff
@@ -35,78 +33,9 @@ def __init__(self):
         self.resourceName = 'oauth'
 
         self.route('GET', ('provider',), self.listProviders)
-        self.route('GET', ('google', 'callback'), self.googleCallback)
+        self.route('GET', (':provider', 'callback'), self.callback)
 
-    @access.public
-    def listProviders(self, params):
-        """
-        TODO Once we have multiple providers, this list should be dynamically
-        built based on the selection of which providers are enabled on the
-        plugin config page.
-        """
-        self.requireParams(('redirect',), params)
-        return {
-            'Google': self._getGoogleUrl(params['redirect'])
-        }
-    listProviders.description = (
-        Description('Get the set of supported OAuth providers and their URLs.')
-        .notes('Will be returned as an object mapping provider name to the '
-               'corresponding URL to direct the user agent to.')
-        .param('redirect', 'Where the user should be redirected upon completion'
-               ' of the OAuth flow.'))
-
-    def _getGoogleUrl(self, redirect):
-        clientId = self.model('setting').get(
-            constants.PluginSettings.GOOGLE_CLIENT_ID)
-
-        if clientId is None:
-            raise Exception('No Google client ID setting is present.')
-
-        callbackUrl = '/'.join((
-            os.path.dirname(cherrypy.url()), 'google', 'callback'))
-
-        query = urllib.parse.urlencode({
-            'response_type': 'code',
-            'access_type': 'online',
-            'client_id': clientId,
-            'redirect_uri': callbackUrl,
-            'state': redirect,
-            'scope': ' '.join(constants.GOOGLE_SCOPES)
-        })
-
-        self._setCsrfToken(redirect)
-
-        return '{}?{}'.format(constants.GOOGLE_AUTH_URL, query)
-
-    @access.public
-    def googleCallback(self, params):
-        self.requireParams(('state', 'code'), params)
-
-        redirect = params['state']
-        self._validateCsrfToken(redirect)
-
-        if 'error' in params:
-            raise cherrypy.HTTPRedirect(redirect)
-
-        clientId = self.model('setting').get(
-            constants.PluginSettings.GOOGLE_CLIENT_ID)
-        clientSecret = self.model('setting').get(
-            constants.PluginSettings.GOOGLE_CLIENT_SECRET)
-
-        user = providers.Google(
-            clientId, clientSecret, cherrypy.url()).getUser(params['code'])
-
-        self.sendAuthTokenCookie(user)
-
-        raise cherrypy.HTTPRedirect(redirect)
-    googleCallback.description = None
-
-    def _setCsrfToken(self, redirect):
-        """
-        Generate an anti-CSRF token and set it in a cookie for the user. It will
-        be verified upon completion of the OAuth login flow.
-        This can be used consistently across all of the OAuth providers.
-        """
+    def _sendCsrfToken(self, redirect):
         csrfToken = self.model('token').createToken(days=0.25)
```

Comment:
The purpose of the [``state`` query string parameter](https://tools.ietf.org/html/rfc6749#section-4.2.1) is primarily to provide CSRF protection. By declining to validate ``state`` on the 2nd half of the login call and using the cookie instead, we are creating a very minor theoretical vulnerability, but more importantly, we're deviating from the OAuth2 standard.

At a minimum, this should be implemented where the cookie contains the redirect location and ``state`` contains the token.

However, I think it'd be even cleaner to get rid of the temporary cookie entirely, and changing ``state`` to be a concatenated combination of a token + a redirect location. Do you have any objections?